esp-hal/src/lib.rs

```diff
@@ -572,3 +572,47 @@ pub fn init(config: Config) -> Peripherals {
 
     peripherals
 }
+
+/// Asynchronous utilities.
+pub mod asynch {
```

Comment:
Can we put this in `sync` instead of lib.rs?

Reply:
Shoving this into the critical section impls isn't exactly a good fit either, I think.

Reply:
Maybe `asynch.rs` then? I'm just conscious that lib.rs can be a bit of a dumping ground, and I'm probably the most guilty of dumping stuff there :sweat_smile:

Reply:
We can do that but it's not a meaningful change. I also want to explore alternatives like not adding a Waker here but a generic one in embassy, and a mutex impl here, so back to draft for a while I guess :)

Reply:
Well I forgot that the constructor needs the mutex, so we need to keep the wrapper struct... moved the module to a new file, if that's all right.